### DISCUSSION TITLE
How to add own javascript code to a Mozilla Hubs room (live server, without "npm run deploy")?

### DISCUSSION THREAD

**Discussion by emclaren** [mozilla/hubs#3601] [Hubs Cloud Questions] [answered]
Note from Elgin: Question was posted by a user in the bugs over here(https://github.com/mozilla/hubs/issues/3595), so I'm transferring it to discussions so it gets seen

NOTE: This is actually a question. I was forwarded here from https://support.mozilla.org/en-US/questions/1318383 ...

Hi!

We are trying to use Mozilla Hubs in a customer case, but we need to add functionality (javascript logic that manages some objects) to the rooms. How can this be done? I've tried:

    ... to search for an injection point in the Hubs server running in AWS, but cannot find it (how the room web page is created and from which templates, and where the templates are?)
    ... to deploy our custom Hubs server using "npm run deploy", but the server gets broken and cannot load rooms any more...

Ideally I'd just mod an existing, running, live server without deploying it again as the customer might not allow redeploying...

PS. We cannot create a browser plugin as the customer uses Oculus Quests to run Hubs, and on Quest you cannot use your own browser plugins...

Thanks!

PS. I'm not trying to mod hubs.mozilla.com, but our own (or the customer's) instance of Hubs running on AWS.

**Comment by softabilityfi** [comment]
New info: I found that if I mod Amazon S3/new-hubs-assets-xyabc123/hubs/pages/latest/hub.html I can see the changes in the room's html, but it seems that running javascript inside that page is disabled? How can it be enabled?

**Comment by softabilityfi** [comment]
I went and added this line to hub.html on S3:

`<script type="application/javascript" src="https://new-hubs-assets.abcxyz.org/hubs/assets/js/helloWorld.js"/>`

helloWorld.js contains just this line of code:

`console.log("HELLO WORLD from S3!");`

... and indeed when I go to a room it's executed:

![kuva](https://user-images.githubusercontent.com/72250567/102779486-fc319880-439c-11eb-8ae6-90bebcea2cc7.png)

But all I get is a white screen. It seems like having a reference to that helloWorld.js somehow breaks the execution of the html page... Any ideas how to fix this?

**Comment by softabilityfi** [comment]
Argh, stupid user error. It should have been written like this:

`<script type="application/javascript" src="https://new-hubs-assets.abcxyz.org/hubs/assets/js/helloWorld.js"></script>`

It seems that /hubs/pages/latest/hub.html on S3 was the correct place!

**Reply by robinkwilson** [reply]
**This isn't the recommended way of doing this**, please see my answer below.

**Comment by robinkwilson** [comment, marked as answer]
Unfortunately, the next automatic update will remove your changes from the S3 bucket file when it is updated. You'll have to add this again every month when Hubs Cloud updates happen - not ideal.

Instead! I would, add your code to the "Additional HTML/Javascript fields" in your Admin Panel:
1. Go to your Admin Panel
2. Click Server Settings menu
3. Click Advanced tab
4. Add your script tags to: "Extra Room Header HTML"
5. OR, if that didn't work, likely because it needs to load after all the DOM content loaded, add your script in "Extra Room Script" BUT you'll need to adapt the code to be Javascript.

Fields screenshots:
 
Extra Room Header HTML:
![image](https://user-images.githubusercontent.com/29560735/102800800-38e39c80-4369-11eb-9877-2bc4369075fd.png)

Extra Room Script
![image](https://user-images.githubusercontent.com/29560735/102800827-44cf5e80-4369-11eb-9930-8cd326be9a0f.png)

**Reply by robinkwilson** [reply]
Marking as answered, please "unmark" and add more information if we need to revisit.
